### PR TITLE
Energy minimization infrastructure

### PR DESCRIPTION
**Pull request by shinzlet**
This PR implements a mock energy minimization algorithm that drives bonds to contract towards a target length and causes unbonded atoms to repel. More importantly, it implements the infrastructure needed to store and operate on energy-minimized molecule structures:

- Energy minimization is applied after every feature
- A molecule's energy-minimized structure is now saved in a checkpoint system
- When the history step is changed, molecules are rebuilt starting at the nearest available checkpoint (rather than being rebuilt from scratch)
- When a molecule is serialized, its current state is saved as a checkpoint

Note that this PR is branched off of #27, which implements the serialization groundwork. I propose closing #27 and merging this PR instead.

Please let me know if any changes should be made. I'm aware that the `scene` crate is starting to balloon, and #26 is still unaddressed - once this is merged I will create a cleanup & refactor PR.

### CODE REVIEW

**Comment by maaku** [comment]
I merged #27

This PR needs to get rebased, and the history is getting pretty messy. Can you do a clean rebase on top of main?

File: crates/scene/feature.md
Comment (maaku) [inline crates/scene/feature.md:1]: Why did you delete this?

Comment (shinzlet) [reply crates/scene/feature.md:1]: It's actually a verbatim copy of https://github.com/shinzlet/atomCAD/issues/5 that I made on my laptop so that I could have it for reference while coding offline. I didn't mean to commit it in the first place - if you read through, you'll notice that parts of it are incomplete and very unclear, as it was just for my own reference.

Rest assured that I'll include all the relevant bits in the real project documentation :)

**Comment by shinzlet** [comment]
Rebased :)
